src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -243,6 +243,35 @@ struct md_mmap_info {
 	size_t  size;
 };
 
+struct md_apk_mmap_info {
+	md_mmap_info mmap_info;
+	md_mmap_info file_info;
+};
+
+MONO_API int monodroid_getpagesize (void);
+
+static md_apk_mmap_info
+md_mmap_apk_file(int fd, uLong offset, uLong size, const char* filename, const char* apk)
+{
+	md_apk_mmap_info info;
+	
+	int pageSize = monodroid_getpagesize();
+	uLong offsetFromPage = offset % pageSize;
+	uLong offsetPage = offset - offsetFromPage;
+	uLong offsetSize = size + offsetFromPage;
+	
+	info.mmap_info.area = mmap(NULL, offsetSize, PROT_READ, MAP_PRIVATE, fd, offsetPage);
```

Comment:
Can/should we also use `MAP_NORESERVE`?  Quoting [the docs](https://linux.die.net/man/2/mmap):

> **MAP_NORESERVE**
> Do not reserve swap space for this mapping. When swap space is reserved, one has the guarantee that it is possible to modify the mapping. When swap space is not reserved one might get **SIGSEGV** upon a write if no physical memory is available. See also the discussion of the file */proc/sys/vm/overcommit_memory* in [**proc**](https://linux.die.net/man/5/proc)(5). In kernels before 2.6, this flag only had effect for private writable mappings.

This is plausibly moot for *Android*, as Android doesn't have swap, but this code could also be used by desktop apps (e.g. the Android designer), which *could* run on Linux...

However, macOS doesn't have `MAP_NORESERVE`...

Reply:
My 2 cents :

Based on what I can see in the kernel code, this is for the overflow accounting, and based on https://github.com/torvalds/linux/blob/6f0d349d922ba44e4348a17a78ea51b7135965b1/Documentation/vm/overcommit-accounting.rst, we have 

> For a file backed map
SHARED or READ-only - 0 cost (the file is the map not swap)
PRIVATE WRITABLE - size of mapping per instance

Since our mappings are file backed and read-only, it doesn't seem like that flag would make any difference (though, in theory, according to the documentation, it's possible to make an implementation that would need that flag).

I also took a quick look at Mono for an example and I only found one usage in libgc for Solaris on a read+write mapping. Though, again, it's possible in theory to have a kernel that needs it and then Mono would have their read-only mapping taken into account.